src/request_ext.rs

```diff
@@ -0,0 +1,295 @@
+            request: self,
+        }
+    }
+}
+
+
+impl From<Agent> for AgentRef<'static> {
+    fn from(value: Agent) -> Self {
+        AgentRef::Owned(value)
+    }
+}
+
+impl<'a> From<&'a Agent> for AgentRef<'a> {
+    fn from(value: &'a Agent) -> Self {
+        AgentRef::Borrowed(value)
+    }
+}
+
+impl Deref for AgentRef<'_> {
+    type Target = Agent;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            AgentRef::Owned(agent) => agent,
+            AgentRef::Borrowed(agent) => &*agent,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+    use super::*;
+    use crate::config::RequestLevelConfig;
+
+    #[test]
+    fn set_https_only_to_true_on_get_request_with_default_agent() {
+        // Create `http` crate request and configure with trait
+        let request = http::Request::builder()
+            .method(http::Method::GET)
+            .uri("http://foo.bar")
+            .body(())
+            .unwrap()
+            .with_default_agent()
+            .configure()
+            .https_only(true)
+            .build();
+
+        // Assert that the request-level configuration has been set
+        let request_config = request
+            .request
+            .extensions()
+            .get::<RequestLevelConfig>()
+            .cloned()
+            .unwrap();
+
+        assert_eq!(request_config.0.https_only(), true);
+    }
+
+    #[test]
+    fn set_https_only_to_false_on_get_request_with_default_agent() {
+        // Create `http` crate request and configure with trait
+        let request = http::Request::builder()
+            .method(http::Method::GET)
+            .uri("http://foo.bar")
+            .body(())
+            .unwrap()
+            .with_default_agent()
+            .configure()
+            .https_only(true)
+            .build();
+
+        // Assert that the request-level configuration has been set
+        let request_config = request
+            .request
+            .extensions()
+            .get::<RequestLevelConfig>()
+            .cloned()
+            .unwrap();
+
+        assert_eq!(request_config.0.https_only(), false);
+    }
+
+    #[test]
+    fn set_http_status_as_error_to_true_on_post_request_with_default_agent() {
+        // Create `http` crate request
+        let request = http::Request::builder()
+            .method(http::Method::POST)
+            .uri("http://foo.bar")
+            .body("Some body")
+            .unwrap();
+
+        // Configure with the trait
+        let request = request.with_default_agent().configure().http_status_as_error(true).build();
+
+        let request_config = request
+            .request
+            .extensions()
+            .get::<RequestLevelConfig>()
+            .cloned()
+            .unwrap();
+
+        assert_eq!(request_config.0.http_status_as_error(), true);
+    }
+
+    #[test]
+    fn set_http_status_as_error_to_false_on_post_request_with_default_agent() {
```

Comment:
These tests all test the `build` behaviour. Optimally I'd also like to test the `with_agent().configure()...run()` behaviour, but I am unsure how.